xdelta3-dir-patcher

```diff
@@ -137,29 +546,53 @@ class XDelta3DirPatcher(object):
 
             self.copy_attributes(patch_path, target_path)
 
+            remove(patch_path)
+
     # TODO: Unit test me
-    def diff(self, old_dir, new_dir, patch_bundle, metadata = None):
-        # Create a temp dir
-        target_dir = mkdtemp(prefix="%s_" % self.__class__.__name__)
+    def diff(self, old_dir, new_dir, patch_bundle, metadata = None,
+             staging_dir = None,
+             runner = ExecutorRunner()):
+        target_dir = mkdtemp(prefix='%s_target' % XDelta3DirPatcher.__name__,
+                             dir=staging_dir)
+
         print("Using %s as staging area" % target_dir)
+        stdout.flush()
 
         delta_target_dir = path.join(target_dir, self.PATCH_FOLDER)
-        mkdir(delta_target_dir)
+        if not path.isdir(delta_target_dir):
+            mkdir(delta_target_dir)
+
+        with XDeltaArchive(old_dir) as old_archive_obj, \
+             XDeltaArchive(new_dir) as new_archive_obj:
+            old_staging_dir = mkdtemp(prefix='%s_old_src' % XDelta3DirPatcher.__name__,
+                                      dir=staging_dir)
+            new_staging_dir = mkdtemp(prefix='%s_new_src' % XDelta3DirPatcher.__name__,
+                                      dir=staging_dir)
 
-        for root, new_dirs, new_files in walk(new_dir):
-            print("Entering directory", root)
+            # XXX: Preload old archive file list to avoid concurrency issues
+            # TODO: Make the listing thread-safe
+            old_archive_obj.list_files()
 
-            rel_path = path.relpath(root, new_dir)
+            for filename in new_archive_obj.list_files():
+                if self.args.debug:
+                    print('Processing %s' % filename)
+                else:
+                    print('.', end = "")
 
-            if self.args.debug: print("- Subdirs:", new_dirs)
-            if self.args.debug: print("- Files:", new_files)
+                runner.add_task(self._find_file_delta, (filename,
+                                                        old_archive_obj,
+                                                        new_archive_obj,
+                                                        old_staging_dir,
+                                                        new_staging_dir,
+                                                        delta_target_dir))
```

Comment:
Like in the applying case, I think this needs more care with directories. The main thing I'm worried about is mode and ownership of directories. Suppose the archive has a read only directory. Once the directory mode is changed on expansion, you won't be able to write any more files into it.

I think instead, you need to start with the files, creating leading directories with normal permissions as needed. After the files are completed, go back through on the directories, creating any that don't exist if necessary and changing the permissions/ownership. In general, I think this needs to start with the leaves and work back towards the root. I'm pretty sure that's how other archiving programs work.